### PR TITLE
Force mr approval

### PR DESCRIPTION
**Pull request by majk-p**
Resolves https://github.com/pitgull/pitgull/issues/120 

Currently targeting pg-43

### CODE REVIEW

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala [hunk 1]
```diff
@@ -166,10 +167,22 @@ object Gitlab {
...
+
+      private def setMrRuleApprovals(projectId: Long, mergeRequestIid: Long, ruleId: Long, amount: Long): F[Unit] =
+        runInfallibleEndpoint(GitlabEndpoints.setMRRuleApprovalRequirement)
+          .apply((projectId, mergeRequestIid, ruleId, amount))
+
       def forceApprove(projectId: Long, mergeRequestIid: Long): F[Unit] =
-        runInfallibleEndpoint(GitlabEndpoints.setApprovalRequirement)
-          .apply((projectId, mergeRequestIid, 0))
-          .void
+        listMRApprovalRules(projectId, mergeRequestIid)
```
Comment (kubukoz) [inline gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:180]: let's use an fs2 stream and run this in parallel for all the PRs. It'll be something roughly like

```scala
Stream.evals(listMRApprovalRules(...)).filter(...)
  .map { rule =>
    setMrRuleApprovals(...)
  }.parJoinUnbounded.compile.drain
```

In the future we can add a global semaphore around the sttp client to ensure we're not spamming the instance :)

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala [hunk 2]
```diff
@@ -204,4 +217,32 @@ object GitlabEndpoints {
       .in("approvals")
       .in(query[Long]("approvals_required"))
 
+  import io.circe.generic.semiauto._
+  import sttp.tapir._
+  import sttp.tapir.json.circe._
+  import io.circe.{ Codec => CirceCodec }
+  
+  final case class ApprovalRule(id: Long, name: String, rule_type: String)
```
Comment (kubukoz) [inline gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:225]: I think we have a circe config somewhere that uses snake_case - it should be reused here

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala [hunk 3]
```diff
@@ -204,4 +217,32 @@ object GitlabEndpoints {
       .in("approvals")
       .in(query[Long]("approvals_required"))
 
+  import io.circe.generic.semiauto._
```
Comment (kubukoz) [inline gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:220]: generic-extras

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 1 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:180]: How about:
```scala
Stream
  .evals(listMRApprovalRules(projectId, mergeRequestIid))
  .filterNot(_.rule_type == "code_owner")
  .evalMap { rule =>
    setMrRuleApprovals(projectId, mergeRequestIid, rule.id, 0)
  }
  .compile
  .drain
```

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 2 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:225]: Nope, I can create dedicated configuration here, though I'm not sure it's worth it for a single class. Which one do you prefer?

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 3 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:220]: Not as long as we don't have configuration. See https://github.com/pitgull/pitgull/pull/123#discussion_r560958268

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 1 above)
Comment (kubukoz) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:180]: That'll be sequential, sure, we can start with that ;)

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 2 above)
Comment (kubukoz) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:225]: Let's put it here then - we will probably need to call the API e.g. for getting the dhall config for a given project, so it'll eventually be reused

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 3 above)
Comment (kubukoz) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:220]: :)

**Comment by kubukoz** [comment]
umm why did this even close? I blame refined-github...

**Comment by kubukoz** [comment]
Back on track :D

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala [hunk 4]
```diff
@@ -169,6 +170,24 @@ object Gitlab {
...
+      private def listMRApprovalRules(projectId: Long, mergeRequestIid: Long): F[List[ApprovalRule]] =
+        runInfallibleEndpoint(GitlabEndpoints.listMRApprovaRules)
+          .apply((projectId, mergeRequestIid))
+
+      private def setMrRuleApprovals(projectId: Long, mergeRequestIid: Long, ruleId: Long, amount: Long): F[Unit] =
+        runInfallibleEndpoint(GitlabEndpoints.setMRRuleApprovalRequirement)
+          .apply((projectId, mergeRequestIid, ruleId, amount))
+
+      def forceApprove(projectId: Long, mergeRequestIid: Long): F[Unit] =
+        listMRApprovalRules(projectId, mergeRequestIid).flatMap { approvalRules =>
```
Comment (kubukoz) [inline gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:183]: I think the comment disappeared, but let's use fs2 here and I'm good to merge

Comment (kubukoz) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:183]: ah, it's still there but on the main view of the review only.

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 1 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala]: done

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 3 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala]: using circe generic extras now

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 2 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala]: done, as for now the config is in `GitlabEndpoints.transport`, we can move it somewhere more generic later

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 4 above)
Comment (majk-p) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala]: using `fs2` now :wink:

**Comment by majk-p** [comment]
Ups, it seems I've missed some of your sttp3 compatibility changes and forced push :sweat_smile:

**Comment by majk-p** [comment]
Issues sorted out, tested manually and confirmed to work well

File: src/main/scala/io/pg/config/format.scala [hunk 5]
```diff
@@ -10,7 +10,7 @@ import io.circe.Decoder
 import io.circe.Encoder
 import io.circe.DecodingFailure
 
-private object circe {
+object circe {
```
Comment (kubukoz) [inline src/main/scala/io/pg/config/format.scala:13]: Now I'm thinking we should merge this one with the config in endpoints... But we can do that later.

**Review by kubukoz — APPROVED** [review]

**Pull request merged**